ag-master/main.py

```diff
@@ -0,0 +1,447 @@
+        existing.file = file
+        existing.command = command
+        db.session.commit()
+        return dict(assign_id=existing.ag_key)
+
+    id = gen_salt(24)
+    existing = Assignment(
+        name=name, course=course.secret, file=file, command=command, ag_key=id
+    )
+    db.session.add(existing)
+    db.session.commit()
+
+    return dict(assign_id=id)
+
+
+@app.route("/jobs/<job>")
+@admin_only
+def job_info(course, job):
+    job = Job.query.filter_by(job_key=job).first()
+    if not job:
+        abort(404)
+
+    assignment = Assignment.query.filter_by(ag_key=job.assignment).first()
+    if not assignment or assignment.course != course.secret:
+        abort(403)
+
+    return {
+        "assignment": assignment.name,
+        "backup": job.backup,
+        "status": job.status,
+        "result": job.result,
+    }
+
+
+@app.route("/course_info")
+@admin_only
+def course_info(course):
+    assignments = Assignment.query.filter_by(course=course.secret)
+    return {
+        "assignments": [
+            {
+                "name": a.name,
+                "file": a.file,
+                "command": a.command,
+                "ag_key": a.ag_key,
+            }
+            for a in assignments
+        ]
+    }
+
+
+########################################
+## Okpy Endpoints
+########################################
+
+
+@app.route("/api/ok/v3/grade/batch", methods=["POST"])
+def batch_grade():
+    data = request.json
+    subms = data["subm_ids"]
+    assignment_id = data["assignment"]
+    if assignment_id == "test":
+        return "OK"
+    ok_token = data["access_token"]
+
+    assignment = Assignment.query.filter_by(ag_key=assignment_id).first()
+    if not assignment:
+        abort(404, "Unknown Assignment")
+
+    jobs = [gen_salt(24) for _ in subms]
+
+    objects = [
+        Job(
+            assignment=assignment.ag_key,
+            backup=id,
+            status="queued",
+            job_key=job_id,
+            access_token=ok_token,
+        )
+        for id, job_id in zip(subms, jobs)
+    ]
+    db.session.bulk_save_objects(objects)
+    db.session.commit()
+
+    try:
+        requests.post(
+            f"{MASTER_URL}/trigger_jobs",
+            json=dict(
+                assignment_id=assignment_id,
+                subms=subms,
+                jobs=jobs,
+            ),
+            headers=dict(Authorization=get_secret(secret_name="AG_MASTER_SECRET")),
+            timeout=1,
+        )
+    except requests.exceptions.ReadTimeout:
+        pass
+
+    return dict(jobs=jobs)
+
+
+@app.route("/trigger_jobs", methods=["POST"])
+@check_master
+def trigger_jobs():
+    data = request.json
+    assignment_id = data["assignment_id"]
+    subms = data["subms"]
+    jobs = data["jobs"]
+
+    assignment = Assignment.query.filter_by(ag_key=assignment_id).first()
+    if not assignment:
+        abort(404, "Unknown Assignment")
+
+    subm_batches = [subms[i : i + 100] for i in range(0, len(subms), 100)]
```

Comment:
100 should be a constant at the top of the file, so we know everywhere it's used